### PR TITLE
Add feature extension

### PR DESCRIPTION
**Pull request by xxbillouxx**

### CODE REVIEW

**Comment by xxbillouxx** [comment]
recheck

**Comment by xxbillouxx** [comment]
recheck

**Comment by fblackburn1** [comment]
recheck

**Comment by xxbillouxx** [comment]
recheck

**Review by pc-m — CHANGES_REQUESTED** [review]
Pretty impressed by all the work you've done for this refactor. Good job

File: xivo_dao/resources/feature_extension/database.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 The Wazo Authors  (see the AUTHORS file)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from xivo_dao.resources.feature_extension.model import ForwardFeatureExtension, \
+    ServiceFeatureExtension, AgentActionFeatureExtension
+
+
+def clean_exten(exten):
+    return exten.strip('_.')
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/resources/feature_extension/database.py]: @pc-m better version used now

File: xivo_dao/resources/func_key/hint_dao.py [hunk 2]
```diff
@@ -37,11 +38,9 @@
 from xivo_dao.resources.func_key.model import Hint
 
 
-def _find_extenfeatures(session, typeval):
-    return session.query(Extension.exten).filter(and_(
-        Extension.context == 'xivo-features',
-        Extension.type == 'extenfeatures',
-        Extension.typeval == typeval,
+def _find_extenfeatures(session, feature):
+    return session.query(FeatureExtension.exten).filter(and_(
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/resources/func_key/hint_dao.py]: @pc-m `and_` removed

File: xivo_dao/alchemy/extension.py [hunk 3]
```diff
@@ -65,51 +68,65 @@ def context_type(self):
...
-            select([Context.contexttype]).where(Context.name == cls.context).label('type'),
+            select([Context.contexttype])
+            .where(Context.name == cls.context)
+            .label('type'),
             'internal',
         )
 
-    dialpattern = relationship('DialPattern',
-                               primaryjoin="""and_(Extension.type == 'outcall',
+    dialpattern = relationship(
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/alchemy/extension.py:77]: @pc-m reformated now

File: xivo_dao/alchemy/feature_extension.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class FeatureExtension(Base):
+    __tablename__ = 'feature_extension'
+    __table_args__ = (
+        PrimaryKeyConstraint('uuid'),
+        UniqueConstraint('exten'),
+        Index('feature_extension__idx__exten', 'exten'),
+        Index('feature_extension__idx__feature', 'feature'),
+        Index('feature_extension__idx__uuid', 'uuid'),
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/alchemy/feature_extension.py]: @pc-m explicit index for uuid removed now

File: xivo_dao/alchemy/feature_extension.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    uuid = Column(UUID(as_uuid=True), server_default=text('uuid_generate_v4()'))
+    enabled = Column(Boolean, nullable=False, server_default='true')
+    exten = Column(String(40), nullable=False)
+    feature = Column(String(255), nullable=False)
+
+    def is_pattern(self):
+        return self.exten.startswith('_')
+
+    def clean_exten(self):
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/alchemy/feature_extension.py]: @pc-m `clean_exten` implementation removed and `clean_extension` used instead

File: xivo_dao/alchemy/tests/test_feature_extension.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+class TestIsPattern(unittest.TestCase):
+
+    def test_is_not_pattern(self):
+        extension = FeatureExtension(exten='1000')
+        assert_that(extension.is_pattern(), equal_to(False))
+
+    def test_is_pattern(self):
+        extension = FeatureExtension(exten='_XXXX')
+        assert_that(extension.is_pattern(), equal_to(True))
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/alchemy/tests/test_feature_extension.py:21]: @pc-m as you suggested, `clean_extension` from `xivo-lib-python` used

File: xivo_dao/resources/feature_extension/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+# Copyright 2023 The Wazo Authors  (see the AUTHORS file)
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/resources/feature_extension/__init__.py]: @pc-m `__init__.py` files cleaned now

File: xivo_dao/resources/feature_extension/model.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2023 The Wazo Authors  (see the AUTHORS file)
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+from typing import NamedTuple
+
+
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/resources/feature_extension/model.py]: @pc-m content moved into `database.py` as used only here

File: xivo_dao/resources/func_key/hint_dao.py [hunk 9]
```diff
@@ -37,31 +40,45 @@
...
-    return session.query(Extension.exten).filter(and_(
-        Extension.context == 'xivo-features',
-        Extension.type == 'extenfeatures',
-        Extension.typeval == typeval,
-    )).scalar()
+def _find_extenfeatures(session, feature):
+    return (
+        session.query(FeatureExtension.exten)
+        .filter(
+            and_(
```
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/resources/func_key/hint_dao.py]: @pc-m `and_` removed now as suggested

File: xivo_dao/alchemy/extension.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pc-m) [reply xivo_dao/alchemy/extension.py:77]: This change needs to be applied to other relationships in this file as well ;)

**Review by pc-m — CHANGES_REQUESTED** [review]

File: xivo_dao/alchemy/extension.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xxbillouxx) [reply xivo_dao/alchemy/extension.py:77]: @pc-m fixed

**Review by pc-m — APPROVED** [review]

**Comment by pc-m** [comment]
recheck

**Pull request merged**